### PR TITLE
Loader reuse existing nodes on reload

### LINKED ISSUE #6566: 0.40.x: import component reports unknown after soft-reload
### What's wrong?

Hi,

I figure out that that the import components remains `unknown` after do a /-/reload via web.

I have the feeling, that `import.http` components reporting `unknown`, because we have 1 import.http at root scope and 3 import.http at `all.config.this`.

Sadly, the `import.http` component are invisible at UI, this take we around a hour the identify what could be root cause to provide a minimal configuration.

### Steps to reproduce

```
jok@CDT-MB-20200178 cdt-ops-stack % curl -s localhost:12345/metrics | grep running
# HELP agent_component_controller_running_components Total number of running components.
# TYPE agent_component_controller_running_components gauge
agent_component_controller_running_components{controller_id="",health_type="healthy"} 2
jok@CDT-MB-20200178 cdt-ops-stack % curl -s localhost:12345/-/reload              
config reloaded
jok@CDT-MB-20200178 cdt-ops-stack % curl -s localhost:12345/metrics | grep running
# HELP agent_component_controller_running_components Total number of running components.
# TYPE agent_component_controller_running_components gauge
agent_component_controller_running_components{controller_id="",health_type="healthy"} 1
agent_component_controller_running_components{controller_id="",health_type="unknown"} 1
```

### System information

Windows / Linux

### Software version

v0.40.0

### Configuration

```text
import.string "a" {
  content = `declare "b" { export "b" { value = "3" } }`
}

a.b "this" {

}
```

### Logs

```text
ts=2024-02-29T17:37:04.191429Z level=info "boringcrypto enabled"=false
ts=2024-02-29T17:37:04.191455Z level=info msg="running usage stats reporter"
ts=2024-02-29T17:37:04.191458Z level=info msg="starting complete graph evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08
ts=2024-02-29T17:37:04.191468Z level=info msg="applying non-TLS config to HTTP server" service=http
ts=2024-02-29T17:37:04.191471Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=http duration=12.5µs
ts=2024-02-29T17:37:04.191475Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=ui duration=250ns
ts=2024-02-29T17:37:04.191478Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=cluster duration=208ns
ts=2024-02-29T17:37:04.19148Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=tracing duration=2.5µs
ts=2024-02-29T17:37:04.191516Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=remotecfg duration=19.75µs
ts=2024-02-29T17:37:04.191519Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=otel duration=208ns
ts=2024-02-29T17:37:04.191523Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=logging duration=94.625µs
ts=2024-02-29T17:37:04.191545Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=import.string.a duration=18.959µs
ts=2024-02-29T17:37:04.191581Z level=info msg="starting complete graph evaluation" controller_id=a.b.this trace_id=7535aaa8a3d7efc24e991be02b8f70cc
ts=2024-02-29T17:37:04.191595Z level=info msg="finished node evaluation" controller_id=a.b.this trace_id=7535aaa8a3d7efc24e991be02b8f70cc node_id=export.b duration=5.833µs
ts=2024-02-29T17:37:04.191601Z level=info msg="finished complete graph evaluation" controller_id=a.b.this trace_id=7535aaa8a3d7efc24e991be02b8f70cc duration=28.875µs
ts=2024-02-29T17:37:04.191738Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=a.b.this duration=187.041µs
ts=2024-02-29T17:37:04.191754Z level=info msg="finished node evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 node_id=labelstore duration=4µs
ts=2024-02-29T17:37:04.191761Z level=info msg="finished complete graph evaluation" controller_id="" trace_id=8adb1639947591d08e59d1900f9afa08 duration=424.166µs
ts=2024-02-29T17:37:04.19204Z level=info msg="scheduling loaded components and services"
ts=2024-02-29T17:37:04.192098Z level=info msg="starting complete graph evaluation" controller_id=a.b.this trace_id=d7df04007f7b974867912defccf329ca
ts=2024-02-29T17:37:04.19211Z level=info msg="finished node evaluation" controller_id=a.b.this trace_id=d7df04007f7b974867912defccf329ca node_id=export.b duration=4µs
ts=2024-02-29T17:37:04.192117Z level=info msg="finished complete graph evaluation" controller_id=a.b.this trace_id=d7df04007f7b974867912defccf329ca duration=48.459µs
ts=2024-02-29T17:37:04.192123Z level=info msg="finished node evaluation" controller_id="" node_id=a.b.this duration=71.916µs
ts=2024-02-29T17:37:04.192154Z level=info msg="scheduling loaded components and services"
ts=2024-02-29T17:37:04.192172Z level=info msg="starting cluster node" peers="" advertise_addr=127.0.0.1:12345
ts=2024-02-29T17:37:04.192286Z level=info msg="peers changed" new_peers=CDT-MB-20200178.fritz.box
ts=2024-02-29T17:37:04.192774Z level=info msg="now listening for http traffic" service=http addr=127.0.0.1:12345
ts=2024-02-29T17:37:16.190953Z level=info msg="reload requested via /-/reload endpoint" service=http
ts=2024-02-29T17:37:16.191205Z level=info msg="starting complete graph evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5
ts=2024-02-29T17:37:16.191232Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=otel duration=4.083µs
ts=2024-02-29T17:37:16.191253Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=http duration=8.708µs
ts=2024-02-29T17:37:16.191265Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=ui duration=1.875µs
ts=2024-02-29T17:37:16.191277Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=cluster duration=3.75µs
ts=2024-02-29T17:37:16.191288Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=logging duration=3.458µs
ts=2024-02-29T17:37:16.191336Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=import.string.a duration=39µs
ts=2024-02-29T17:37:16.19138Z level=info msg="starting complete graph evaluation" controller_id=a.b.this trace_id=87437568596d7102ae216a96fda87dc4
ts=2024-02-29T17:37:16.19146Z level=info msg="finished node evaluation" controller_id=a.b.this trace_id=87437568596d7102ae216a96fda87dc4 node_id=export.b duration=6.417µs
ts=2024-02-29T17:37:16.191476Z level=info msg="finished complete graph evaluation" controller_id=a.b.this trace_id=87437568596d7102ae216a96fda87dc4 duration=112.417µs
ts=2024-02-29T17:37:16.191491Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=a.b.this duration=143.833µs
ts=2024-02-29T17:37:16.191541Z level=info msg="scheduling loaded components and services"
ts=2024-02-29T17:37:16.191744Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=remotecfg duration=243.417µs
ts=2024-02-29T17:37:16.191765Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=labelstore duration=4.25µs
ts=2024-02-29T17:37:16.19178Z level=info msg="finished node evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 node_id=tracing duration=6.625µs
ts=2024-02-29T17:37:16.191788Z level=info msg="finished complete graph evaluation" controller_id="" trace_id=65f9ded278648d30fc2dc17571aa71c5 duration=661.916µs
ts=2024-02-29T17:37:16.191803Z level=info msg="config reloaded" service=http
ts=2024-02-29T17:37:16.191851Z level=info msg="scheduling loaded components and services"
ts=2024-02-29T17:37:16.191972Z level=info msg="starting complete graph evaluation" controller_id=a.b.this trace_id=a63873984d2b42f8ffe4b2192da95728
ts=2024-02-29T17:37:16.192006Z level=info msg="finished node evaluation" controller_id=a.b.this trace_id=a63873984d2b42f8ffe4b2192da95728 node_id=export.b duration=17.708µs
ts=2024-02-29T17:37:16.192024Z level=info msg="finished complete graph evaluation" controller_id=a.b.this trace_id=a63873984d2b42f8ffe4b2192da95728 duration=103.875µs
ts=2024-02-29T17:37:16.192039Z level=info msg="finished node evaluation" controller_id="" node_id=a.b.this duration=160µs
ts=2024-02-29T17:37:16.19205Z level=info msg="scheduling loaded components and services"
```

### PR DESCRIPTION
**Pull request by wildum**
#### PR Description

This change optimizes the loader to avoid creating new nodes (declare + config) when they already exist in the graph. This optimization was already implemented for the componentNodes, this extends it to configNodes and declare nodes.

This also fixes a bug where the import nodes where not run after a reload.
Fixes #6566

### CODE REVIEW

File: pkg/flow/internal/controller/node_config_logging.go [hunk 1]
```diff
@@ -80,3 +81,19 @@ func (cn *LoggingConfigNode) Block() *ast.BlockStmt {
...
+// UpdateBlock updates the River block used to construct arguments.
+// The new block isn't used until the next time Evaluate is invoked.
+//
+// UpdateBlock will panic if the block does not match the component ID of the
+// LoggingConfigNode.
+func (cn *LoggingConfigNode) UpdateBlock(b *ast.BlockStmt) {
+	if !BlockComponentID(b).Equals(strings.Split(cn.nodeID, ".")) {
+		panic("UpdateBlock called with an River block with a different ID")
+	}
+
```
Comment (wildum) [inline pkg/flow/internal/controller/node_config_logging.go:94]: if https://github.com/grafana/agent/pull/5992 is merged first, we might need to re-create the buffer here

File: pkg/flow/internal/controller/loader.go [hunk 2]
```diff
@@ -386,19 +386,31 @@ func (l *Loader) populateConfigBlockNodes(args map[string]any, g *dag.Graph, con
...
+				diags.Add(diag.Diagnostic{
+					Severity: diag.SeverityLevelError,
+					Message:  fmt.Sprintf("%q block already declared at %s", node.NodeID(), configBlockStartPos),
+					StartPos: configBlockStartPos,
+					EndPos:   ast.EndPos(block).Position(),
+				})
 
-			continue
+				continue
+			}
```
Comment (rfratto) [inline pkg/flow/internal/controller/loader.go:413]: I think moving this check here means that we won't report on duplicate config blocks anymore.

```river 
logging {} 
logging {} // should error
```

Can you add a test for this (unfortunately it seems like we don't already have one) and make sure it errors?

You'll notice that `populateComponentNodes` creates a temporary map for asserting uniqueness.

**Review by rfratto — COMMENTED** [review]
Mostly looks LGTM but I found something that _might_ be a regression. Once we have a test for it and fix the regression (if there is one), I'll approve and this will be good to go.

File: pkg/flow/internal/controller/loader.go [hunk 3]
```diff
@@ -530,6 +545,21 @@ func (l *Loader) populateComponentNodes(g *dag.Graph, componentBlocks []*ast.Blo
...
+	if orig, redefined := blockMap[id]; redefined {
+		return diag.Diagnostic{
+			Severity: diag.SeverityLevelError,
+			Message:  fmt.Sprintf("block %s already declared at %s", id, ast.StartPos(orig).Position()),
+			StartPos: block.NamePos.Position(),
+			EndPos:   block.NamePos.Add(len(id) - 1).Position(),
+		}, true
+	}
+	blockMap[id] = block
+	return diag.Diagnostic{}, false
```
Comment (wildum) [inline pkg/flow/internal/controller/loader.go:560]: packed the logic in a function for readability and consistency between the populate functions. Should "blockMap[id] = block" be inside of the func? It reduces redundancy but it also feels a bit weird to me that this func modifies the provided map

File: internal/flow/internal/controller/loader.go [hunk 4]
```diff
@@ -539,6 +554,21 @@ func (l *Loader) populateComponentNodes(g *dag.Graph, componentBlocks []*ast.Blo
...
+		return diag.Diagnostic{
+			Severity: diag.SeverityLevelError,
+			Message:  fmt.Sprintf("block %s already declared at %s", id, ast.StartPos(orig).Position()),
+			StartPos: block.NamePos.Position(),
+			EndPos:   block.NamePos.Add(len(id) - 1).Position(),
+		}, true
+	}
+	blockMap[id] = block
+	return diag.Diagnostic{}, false
+}
```
Comment (rfratto) [inline internal/flow/internal/controller/loader.go:570]: nit: I have to scroll pretty far down from where this is first used to where it is defined, which makes it difficult to follow along with the logic. Can you move this function to be directly under its first usage?

File: internal/flow/internal/controller/loader.go [hunk 5]
```diff
@@ -350,16 +356,25 @@ func (l *Loader) populateDeclareNodes(g *dag.Graph, declareBlocks []*ast.BlockSt
...
+			node = exist.(*DeclareNode)
+			node.UpdateBlock(declareBlock)
+		} else {
+			node = NewDeclareNode(declareBlock)
+			if g.GetByID(node.NodeID()) != nil {
+				diags.Add(diag.Diagnostic{
+					Severity: diag.SeverityLevelError,
+					Message:  fmt.Sprintf("cannot add declare node %q; node with same ID already exists", node.NodeID()),
+				})
+				continue
```
Comment (rfratto) [inline internal/flow/internal/controller/loader.go:375]: Is this check required anymore? It seems like the `blockAlreadyDefined` call would handle this case now.

**Review by rfratto — APPROVED** [review]
LGTM % nits

**Pull request merged**